### DISCUSSION TITLE
High latencies when resizing the AnyIO sync thread pool

### DISCUSSION THREAD

**Discussion by bfreuden** [fastapi/fastapi#12269] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import List
import time
from anyio import to_thread
from fastapi import FastAPI
from fastapi.responses import Response
import json

app = FastAPI()

@app.on_event("startup")
def startup():
    limiter = to_thread.current_default_thread_limiter().total_tokens = 1

@app.get("/sync", response_model=List[str])
def get_sync():
    time.sleep(5)
    # return Response(content=json.dumps([]), media_type="application/json")
    return []
```

### Description

I have a sync endpoint and I am limiting the AnyIO sync thread pool to 1 because my app is running in a Docker container with a CPU limit of 1. 
The problem is I'm observing unexpected high latencies , even in simple scenarios.

If you run a single request it is ran in 5 seconds or so (that is expected):
``` 
$ time curl localhost:8000/sync
[]
real	0m5,032s
user	0m0,016s
sys	0m0,013s

```
But if you start 2 requests at almost the same time, they both take 10 seconds or so to complete:
```
 $ time curl localhost:8000/sync
[]
real	0m9,713s
user	0m0,003s
sys	0m0,018s
```
```
$ time curl localhost:8000/sync
[]
real	0m10,035s
user	0m0,017s
sys	0m0,013s
```
And if you start 3 requests at the same time, all will take 15 seconds to complete.

If you uncomment the "return" line, then the first request is returning in 5 seconds and the second request is returning in 10 seconds, which is what would be expected from a user perspective.

It seems to me the validation and encoding of the response of sync endpoints is done on the sync thread pool (it does not seem to be case for async endpoints though, and that came as a surprise to me). Don't get me wrong: it looks like a good approach. However I have the impression that, the way it is implemented, the request is executed in two steps each requiring a worker thread, as if it was first acquiring a semaphore to execute the user code (my `get_sync` function) then releasing the semaphore and reacquiring it again to validate and encode the output (and that's where it get starved).

Note: in a Docker context it is crucial to make sure the container is doing its best to respect its CPU limits otherwise latencies will increase way beyond what would be expected by one's common sense. If you assign 1 CPU core and if the app is trying to run 4 threads at 100%, then latencies will not be increased by 4: they will be increased by more than 10.
That is nicely explained by a Twitter engineer here: https://danluu.com/cgroup-throttling/

### Operating System

Linux

### FastAPI Version

0.115.0

### Pydantic Version

2.9.2

### Python Version

3.12.6

**Comment by bfreuden** [comment]
I'm realizing I forgot to ask the question :)
Do you think that qualifies as a bug? I would say so.
Possible solutions could be making sure to acquire the semaphore only once per http request, or maybe to validate and encode the response on the main event loop thread (since that seems to be acceptable for async endpoints)?

**Comment by YuriiMotov** [comment, marked as answer]
It's not a bug.

FastAPI runs input and output data validation in thread. Your endpoint is also `sync` and will be executed in thread.

Since you limit threadpool capacity to 1, FastAPI will have to queue up all these operations and execute them subsequently in one thread. Since you send 2 requests it will enqueue tasks to execute in threadpool as follows:
* validate the input data of the first request (if there are input parameters)
* validate the input data of the second request (if there are input parameters)
* execute the endpoint for the first request
* execute the endpoint for the second request
* validate the response of the first request
* validate the response of the second request

And, if you uncomment the `return Response(content=json.dumps([]), media_type="application/json")` line, it will skip the response validation step (FastAPI doesn't validate response if you return an instance of `Response` class or its sub-classes).

**Comment by bfreuden** [comment]
Thanks for your answer.
Now I can understand the reason behind that behavior but it's hard to accept that it is not a bug.
Fastapi could queue the operations in this order:

- validate the input data of the first request (if there are input parameters)
- execute the endpoint for the first request
- validate the response of the first request
- validate the input data of the second request (if there are input parameters)
- execute the endpoint for the second request
- validate the response of the second request

And the response of the first request could be sent before the second request is processed.

**Reply by YuriiMotov** [reply]
It's not a bug.
You don't need to limit thread pool capacity to 1 just because you have only 1 CPU. It's totally ok to have 40 threads (and even more) with 1 CPU.
Number of CPUs should be taken into consideration when you decide how many workers you run. It's another thing

**Reply by bfreuden** [reply]
It's far from being totally OK to have more than 1 thread if you have a cgroup limit of 1 CPU and if you program is CPU bound.
Read this: https://danluu.com/cgroup-throttling/
The paper is not from "nobody me", it's from Twitter engineers.

**Reply by YuriiMotov** [reply]
When you run `sync` endpoint in FastAPI you already have 2 threads: main thread, in which the event loop is run, and the thread of threadpool where FASTAPI executes sync code.

Off course, the less threads you have the less will be overheads for switching between them.

If you aim to achieve the highest CPU usage efficiency and it's OK that your server will be able to serve only 1 request at a time, you may consider executing sync code inside an async endpoint. It will block event loop, but there will be not big difference between this approach and the the approach with executing sync endpoints having only 1 thread in thread pool.
It's not good practice, but depending on how heavy you blocking code is, it may work for you. 

If it's possible to split your sync code into blocks, you can use this apptoach:
```python
@app.get('/')
async def func():
    for _ in range(10):
        time.sleep(1)  # imitate the block of sync code
        await asyncio.sleep(0)  # add this to process other tasks in event loop
```
